### PR TITLE
MAE Surface Pretrain: masked autoencoder initialization for backbone

### PR DESCRIPTION
**Pull request by tcapelle**
## Hypothesis

The Transolver backbone (3 blocks, 192 hidden dims) is trained from random initialization end-to-end on CFD label prediction. The backbone must simultaneously learn (a) geometry/physics embedding and (b) flow prediction. Masked Autoencoder (MAE) pretraining forces the backbone to learn spatially coherent geometry embeddings BEFORE seeing any CFD labels.

Pretraining task: mask 50% of surface nodes, reconstruct their 24-dim input features from unmasked context. No CFD labels needed — self-supervised on geometry alone. After pretraining, fine-tune the full model on CFD labels as usual. The MAE head is discarded after pretraining.

**Key bet:** A backbone pretrained on surface geometry reconstruction will have representations that generalize better to OOD geometries (p_oodc) and tandem configurations (p_tan) because it has learned the structure of airfoil surface coordinate manifolds independently of the pressure labels.

**Literature:**
- He et al. "MAE" (arXiv:2111.06377, NeurIPS 2021)
- Pang et al. "Point-MAE" (arXiv:2204.00053, 2022) — MAE for 3D point clouds
- Chen et al. "RI-MAE" (arXiv:2406.11501, 2024) — MAE on irregular grids for PDE surrogates, 8-15% OOD error reduction
- OmniFluids (arXiv:2506.10862, 2025) — physics-only pretrain → finetune for fluid surrogates

### Step 1: Add MAE Components

Add a `MaskedAutoencoderHead` class in train.py:

```python
class MaskedAutoencoderHead(nn.Module):
    def __init__(self, n_hidden=192, in_dim=24):
        super().__init__()
        self.mask_token = nn.Parameter(torch.zeros(1, in_dim))
        nn.init.normal_(self.mask_token, std=0.02)
        self.decoder = nn.Sequential(
            nn.Linear(n_hidden, 128),
            nn.SiLU(),
            nn.Linear(128, in_dim),
        )

    def forward(self, backbone_hidden, mask):
        pred = self.decoder(backbone_hidden)
        return pred, mask
```

### Step 2: Add MAE Pretraining Logic

During MAE pretraining epochs:
1. For each batch, identify surface nodes (`is_surface` or equivalent flag from the data)
2. Randomly mask 50% of surface nodes (volume nodes always visible)
3. Replace masked node features with the learned `mask_token`
4. Forward through backbone ONLY (no SRF head, no PCGrad, no aft-foil SRF)
5. Compute MSE loss between reconstructed features at masked positions and original features
6. Optimize with Lion optimizer at lr=5e-4 for the pretraining phase

### Step 3: Two-Phase Training Protocol

- **Phase 1 (epochs 0 to `mae_pretrain_epochs`):** MAE pretraining only
  - Train backbone + MAE decoder head on surface node reconstruction
  - SRF head parameters frozen / not updated
  - No CFD labels used, no surface MAE computed
  - Faster per-epoch (no PCGrad, no aft-foil SRF forward)

- **Phase 2 (remaining epochs):** Standard CFD fine-tuning
  - Discard MAE decoder head entirely (delete from model)
  - Unfreeze all parameters
  - Train with full CFD pipeline (PCGrad, SRF, aft-foil SRF, DCT loss, etc.)
  - Use standard lr=2e-4 with cosine schedule from this point (T_max=150 covers Phase 2 epochs)
  - EMA starts updating from Phase 2 start, not from epoch 0

### Step 4: New Flags

- `--mae_pretrain_epochs 20` (int, default 0 = disabled) — number of MAE pretraining epochs
- `--mae_mask_ratio 0.5` (float) — fraction of surface nodes to mask
- `--mae_lr 5e-4` (float) — learning rate for MAE pretraining phase

### Step 5: Training Runs

Run 2 seeds:

```
cd cfd_tandemfoil && python train.py --agent askeladd --seed 42 \
  --wandb_name "askeladd/mae-surface-pretrain-s42" \
  --wandb_group "mae-surface-pretrain" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --mae_pretrain_epochs 20 --mae_mask_ratio 0.5 --mae_lr 5e-4
```

Then seed 73 with same flags but `--seed 73`.

### Important Implementation Notes

- Slice assignment in Transolver uses the input features to compute attention. During MAE pretraining, ensure slices are computed from UNMASKED geometric features (coordinates, DSDF) not the masked features. If the slice assignment uses the full input tensor, you may need to pass the original unmasked coordinates separately.
- MAE pretraining epochs are "free" in terms of effective fine-tuning budget — they're faster per-epoch since no PCGrad/SRF is needed. With 20 pretrain epochs, you still get ~128 fine-tune epochs within the 180-min wall clock.
- The cosine_T_max=150 schedule applies to Phase 2 ONLY. Phase 1 uses a flat lr of 5e-4.
- EMA should start updating from Phase 2 (epoch `mae_pretrain_epochs`), not from epoch 0.
- Log MAE reconstruction loss to W&B during Phase 1 for debugging (`mae/recon_loss`).

## Baseline

Current best metrics (PR #2251, 2-seed avg):
- p_in: **11.891** (target: < 11.89)
- p_oodc: **7.561** (target: < 7.56)
- p_tan: **28.118** (target: < 28.12)
- p_re: **6.364** (target: < 6.36)
- W&B baseline runs: 7jix2jkg (s42), epkfhxfl (s73)

### CODE REVIEW

**Comment by tcapelle** [comment]
STUDENT askeladd:

## Results

### Summary

MAE surface pretraining with 20 epochs of hidden-level masking (50% mask ratio) significantly **degraded all metrics** compared to baseline. This is a clear **negative result**.

### Metrics

| Metric | Baseline (2-seed avg) | S42 (mae-pretrain) | S73 (mae-pretrain) | 2-seed avg | Δ vs baseline |
|--------|----------------------|--------------------|--------------------|------------|---------------|
| p_in | **11.891** | 167.2 | 38.4 | 102.8 | **+765%** |
| p_oodc | **7.561** | 139.3 | 43.0 | 91.2 | **+1106%** |
| p_tan | **28.118** | 198.0 | 50.7 | 124.4 | **+342%** |
| p_re | **6.364** | 114.5 | 28.5 | 71.5 | **+1024%** |

### Run Details

**Seed 42** (W&B: `mxtsralg`)
- Best epoch: 60 (of 123 fine-tuning epochs), val_loss = 4.4692
- val_in_dist=3.75, val_tandem=6.54, val_oodc=3.96, val_ood_re=3.63
- Peak VRAM: 46.8 GB

**Seed 73** (W&B: `ngponq4y`)
- Best epoch: 123, val_loss = 1.1654
- val_in_dist=0.87, val_tandem=1.75, val_oodc=1.14, val_ood_re=0.91
- Peak VRAM: 45.8 GB

**MAE Pretraining Phase:**
- 20 epochs, ~30 min wall clock
- Reconstruction loss: 14.6 → 1.36 (S42), 12.2 → 1.41 (S73)
- Fine-tuning: 123 epochs in remaining ~150 min

**Command:**
```
cd cfd_tandemfoil && python train.py --agent askeladd --seed 42 \
  --wandb_name "askeladd/mae-surface-pretrain-s42" \
  --wandb_group "mae-surface-pretrain" \
  --asinh_pressure --asinh_scale 0.75 --field_decoder --adaln_output --use_lion --lr 2e-4 \
  --aug aoa_perturb --aug_full_dsdf_rot --high_p_clamp --n_layers 3 --slice_num 96 \
  --tandem_ramp --domain_layernorm --domain_velhead --ema_decay 0.999 --weight_decay 5e-5 \
  --cosine_T_max 150 --pcgrad_3way --pcgrad_extreme_pct 0.15 \
  --pressure_first --pressure_deep \
  --residual_prediction --surface_refine --surface_refine_hidden 192 --surface_refine_layers 3 \
  --aft_foil_srf --aug_gap_stagger_sigma 0.02 --aug_dsdf2_sigma 0.05 \
  --gap_stagger_spatial_bias \
  --dct_freq_loss --dct_freq_weight 0.05 --dct_freq_gamma 2.0 --dct_freq_alpha 1.5 \
  --te_coord_frame --wake_deficit_feature \
  --mae_pretrain_epochs 20 --mae_mask_ratio 0.5 --mae_lr 5e-4
```

### What happened

The MAE pretraining catastrophically hurt performance. Key observations:

1. **Massive seed sensitivity**: S42 (p_in=167) is 4x worse than S73 (p_in=38), both far worse than baseline (11.9). The MAE initialization creates an extremely unstable starting point that makes fine-tuning outcomes highly seed-dependent.

2. **S42 never converged**: Training loss plateaued at vol~1.3/surf~0.12 (vs S73's vol~0.38/surf~0.03). Best epoch was 60 (early), after which the model quality stagnated. The backbone appears stuck in a bad basin of attraction.

3. **S73 converged but poorly**: Despite healthy-looking learning curves (val_loss steadily improving to 1.17), S73's final metrics are still 3-5x worse than baseline across all splits.

4. **Implementation note**: Masking was done at the hidden embedding level (post-preprocess, pre-transformer-blocks) to ensure slice assignments in attention used unmasked features. The reconstruction target was raw 24-dim input features. This is slightly different from standard MAE which masks at the input/patch level.

### Why it failed

The core issue is that the Transolver backbone is a 3-block, 192-dim model — far smaller than the ViTs where MAE pretraining works well. In the MAE literature, pretraining helps large models (ViT-Base/Large, 12-24 blocks, 768-1024 dim) by preventing overfitting. Our small model doesn't have this overfitting problem, and the pretraining creates an initialization that is optimized for geometry reconstruction rather than flow prediction. The 20 MAE epochs push weights into a reconstruction-optimal region that is far from flow-prediction-optimal, requiring extensive fine-tuning to recover.

Additionally, the hidden-level masking approach may be suboptimal: the model's preprocessing layers (feature_cross, preprocess MLP) see all inputs during MAE, only the transformer blocks see masked embeddings. This creates a mismatch between pre-processing and attention that may confuse the model.

### Suggested follow-ups

1. **Reduce MAE epochs significantly** (e.g., 3-5 instead of 20) to provide a lighter initialization signal without over-committing to reconstruction
2. **Use MAE as an auxiliary loss during regular training** instead of as a separate pretrain phase — add a small reconstruction term to the training objective
3. **Test on a larger backbone** (6+ blocks, 384+ dim) where pretraining benefits are more established
4. **Input-level masking** with proper handling of the preprocessing pipeline, though this requires significant refactoring
5. **Consider this hypothesis closed** — the small model size makes pretraining-then-finetuning fundamentally unsuited to this architecture

**Comment by tcapelle** [comment]
ADVISOR: Closing PR #2276 because ADVISOR: Closing — catastrophic failure, all metrics 3-10x worse than baseline (p_in +765%, p_oodc +1106%, p_tan +342%, p_re +1024%).

**Root cause:** MAE pretraining is designed for large models (ViT-Base/Large, 12-24 blocks, 768-1024 dim). Our 3-block, 192-dim Transolver is too small — it doesn't have the overfitting problem that pretraining solves. 20 MAE epochs push the backbone into a reconstruction-optimal basin that's catastrophically far from flow-prediction-optimal. Even 123 fine-tuning epochs can't recover.

**Additional factors:** Extreme seed sensitivity (4x gap between seeds) confirms unstable initialization. Hidden-level masking creates mismatch between preprocessing (sees all inputs) and attention (sees masked embeddings).

**Key takeaway:** Self-supervised pretraining requires model scale to be useful. Our backbone is small enough that random init + end-to-end training is strictly better. Added to DO NOT REVISIT.

Student suggestion of MAE as auxiliary loss during regular training (not separate phase) is interesting — but the fundamental capacity argument still applies. Moving you to a fresh direction..
